src/operator/random/pdf_op.h

```diff
@@ -0,0 +1,594 @@
+    const DType fmu(mu[index]), falpha(alpha[index]), den(fmu*falpha+1.0);
+
+    // Reparameterize with limit = 1/alpha, prob = 1/(mu*alpha+1)
+    const DType limit(1.0/falpha), prob(1.0/(fmu*falpha+1.0));
+
+    const int end = start + length;
+    for (int i = start; i < end; ++i) {
+        // Grad returned as d_limit, d_prob
+        DType grad_l(0), grad_p(0);
+        PDF_NegativeBinomial_Grad<logpdf>::LPDF_GRAD(limit, prob,
+            DType(sample[i]), out[i],
+            grad_out[i], &grad_l, &grad_p);
+        grad_mu[i]     = -grad_p*falpha/(den*den);
+        grad_alpha[i]  = -grad_l/(falpha*falpha)-grad_p*fmu/(den*den);
+        KERNEL_ASSIGN(grad_sample[i], req, 0);
+    }
+  }
+};
+
+template<bool logpdf>
+struct PDF_Dirichlet {
+  template<typename DType, typename IType1, typename IType2>
+  MSHADOW_XINLINE static void Map(int start, int length, int sample_size, int index, int k,
+                                  DType *out, IType1 *sample, IType2 *alpha) {
+    const int end = start + length;
+    for (int i = start; i < end; ++i) {
+        const IType1 *cur_sample = sample+i*k;
+        const IType2 *cur_alpha  = alpha+index*k;
+        DType sum_alpha(0), sum_lgamma(0), sum_sample(0);
+        for ( int j = 0; j < k; ++j ) {
+          sum_alpha  += cur_alpha[j];
+          sum_lgamma += lgamma(cur_alpha[j]);
+          sum_sample += (cur_alpha[j]-1)*log(cur_sample[j]);
+        }
+        DType lpdf(sum_sample+(lgamma(sum_alpha)-sum_lgamma));
+        out[i] = logpdf ? lpdf : DType(exp(lpdf));
+    }
+  }
+};
+
+
+template<bool logpdf>
+struct PDF_Dirichlet_Grad {
+  template<typename DType, typename IType1, typename IType2>
+  MSHADOW_XINLINE static void Map(int start, int length, int sample_size, int index,
+                  OpReqType req, int k,
+                  DType *out, IType1 *sample, IType2 *alpha,
+                  DType *grad_out, IType1 *grad_sample, IType2 *grad_alpha) {
+    const int end = start + length;
+    for (int i = start; i < end; ++i) {
+        // Digamma function
+        const IType1 *cur_sample = sample+i*k;
+        const IType2 *cur_alpha  = alpha+index*k;
+        const DType scaling(grad_out[i]*(logpdf ? DType(1) : out[i]));
+        DType sum_alpha(0);
+        for ( int j = 0; j < k; ++j ) {
+          sum_alpha += cur_alpha[j];
+        }
+        const DType psi_sum(ceph_psi(sum_alpha));
+        for ( int j = 0; j < k; ++j ) {
+          // order grad_alpha differently to allow efficient reduction at the end.
+          grad_alpha[i%sample_size+sample_size*(j+k*index)] =
+            scaling * (log(cur_sample[j])+(psi_sum-ceph_psi(cur_alpha[j])));
+          KERNEL_ASSIGN(grad_sample[i*k+j], req, scaling*(cur_alpha[j]-1)/cur_sample[j]);
+        }
+    }
+  }
+};
+
+struct PdfParam : public dmlc::Parameter<PdfParam> {
+  bool is_log;
+  DMLC_DECLARE_PARAMETER(PdfParam) {
+    DMLC_DECLARE_FIELD(is_log).set_default(false)
+    .describe("Whether to compute the log PDF or not.");
```

Comment:
May be that can be phrased a bit better.

Reply:
It's now "If set, compute the density of the log-probability instead of the probability."